### PR TITLE
Add core structure changed events

### PR DESCRIPTION
**Pull request by JakaMohorkoDS**
## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] Pull request title reflects its content
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Description:

Second part of the Core Changed events. Now triggers on structural changes. Core object support has been improved to allow for triggers when nested objects change.

### CODE REVIEW

**Comment by JakaMohorkoDS** [comment]
Missing: 
 - Documentation final update
 - Additional trigger for "Component Modified" (eg. when "active" changes)

File: core/opendaq/component/include/opendaq/component_impl.h [hunk 1]
```diff
@@ -439,4 +460,29 @@ std::string ComponentImpl<Intf, Intfs...>::getRelativeGlobalId(const std::string
...
 
+template <class Intf, class ... Intfs>
+void ComponentImpl<Intf, Intfs...>::triggerCoreEvent(const CoreEventArgsPtr& args)
+{
+    try
+    {
+        const ComponentPtr thisPtr = this->template borrowPtr<ComponentPtr>();
+        this->coreEvent(thisPtr, args);
+    }
+    catch (const DaqException& e)
```
Comment (dejancrn) [inline core/opendaq/component/include/opendaq/component_impl.h:471]: I think the handling of std::exception only covers all three cases here

File: core/opendaq/component/include/opendaq/folder_impl.h [hunk 2]
```diff
@@ -163,54 +167,103 @@ ErrCode FolderImpl<Intf, Intfs...>::addItem(IComponent* item)
...
+        if (OPENDAQ_FAILED(err))
+            return err;
+    }
 
-            return OPENDAQ_SUCCESS;
-        });
+    if (!this->coreEventMuted && this->coreEvent.assigned())
+    {
+        const auto component = ComponentPtr::Borrow(item);
+        const auto args = createWithImplementation<ICoreEventArgs, CoreEventArgsImpl>(
```
Comment (dejancrn) [inline core/opendaq/component/include/opendaq/folder_impl.h:189]: What is the reason for not using factories? Doesn't that force that CoreEventArgsImpl has to be header-only implementation? If true this increases compilation time of modules.

**Review by dejancrn — APPROVED** [review]

File: core/opendaq/component/include/opendaq/folder_impl.h (hunk 2 above)
Comment (JakaMohorkoDS) [reply core/opendaq/component/include/opendaq/folder_impl.h:189]: Issues with dependencies. Events are called from within the Component implementation, but the factories require a dependency to signals. So we'd need to have several headers for factory definitions in the opendaq project. 

So I suppose it's a issue of compile time vs. organization clarity.

File: core/coreobjects/include/coreobjects/property_object_impl.h [hunk 3]
```diff
@@ -1696,6 +1726,94 @@ ErrCode GenericPropertyObjectImpl<PropObjInterface, Interfaces...>::checkForRefe
...
+    {
+        if (item.second.assigned())
+        {
+            const auto objInternal = item.second.template asPtrOrNull<IPropertyObjectInternal>();
+            if (objInternal.assigned())
+                objInternal.enableCoreEventTrigger();
+        }
+    }
+
+    for (const auto& item : localProperties)
```
[the remark this answers is not in the document]
Comment (JakaMohorkoDS) [reply core/coreobjects/include/coreobjects/property_object_impl.h:1744]: Up until now property objects that are set as defaults were not frozen. Should that be the case?

File: core/opendaq/component/include/opendaq/component_impl.h [hunk 4]
```diff
@@ -332,10 +341,22 @@ ErrCode INTERFACE_FUNC ComponentImpl<Intf, Intfs...>::update(ISerializedObject*
...
+            if (!muted)
+                propInternalPtr.disableCoreEventTrigger();
+
             const auto err = Super::update(objPtr);
 
             updateObject(objPtr);
 
+            if (!muted && this->coreEvent.assigned())
+            {
+                const CoreEventArgsPtr args = createWithImplementation<ICoreEventArgs, CoreEventArgsImpl>(core_event_ids::ComponentUpdateEnd, Dict<IString, IBaseObject>());
```
[the remark this answers is not in the document]
Comment (JakaMohorkoDS) [reply core/opendaq/component/include/opendaq/component_impl.h:356]: What would nested property objects trigger - ComponentUpdateEnd? PropertyObjectEnd provides a list of updated properties at the moment, and is triggered with begin/end update. ComponentUpdateEnd is called only after `IComponent::update()`

File: core/coreobjects/include/coreobjects/property_object_impl.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (JakaMohorkoDS) [reply core/coreobjects/include/coreobjects/property_object_impl.h:1744]: This will be addressed in a separate PR (jira TBBAS-1213)

**Pull request merged**